File: src/main/java/com/blogly/blogly/application/usecase/CreatePostUseCase.java
package com.blogly.blogly.application.usecase;

import com.blogly.blogly.application.dto.CreatePostRequest;
import com.blogly.blogly.domain.exception.TitleAlreadyExistsException;
import com.blogly.blogly.domain.post.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor
@Component
public class CreatePostUseCase {

    private final PostRepository repository;

    public PostId execute(CreatePostRequest request) {
        var title = new Title(request.title());

        if (repository.existsByTitle(title)) {
            throw new TitleAlreadyExistsException(title.value());
        }

        var post = Post.create(
                title,
                new Content(request.content()));

        return repository.save(post);
    }
}

File: src/main/java/com/blogly/blogly/application/post/CreatePostUseCase.kt
package com.blogly.blogly.application.post

import com.blogly.blogly.application.post.dto.CreatePostRequest
import com.blogly.blogly.application.shared.IdGenerator
import com.blogly.blogly.application.shared.UserProvider
import com.blogly.blogly.domain.post.*
import com.blogly.blogly.domain.post.exception.TitleAlreadyExistsException
import org.springframework.stereotype.Component

@Component
class CreatePostUseCase(
    private val repository: PostRepository,
    private val userProvider: UserProvider,
    private val idGenerator: IdGenerator
) {
    fun execute(request: CreatePostRequest): PostId {
        val title = Title(request.title)

        if (repository.existsByTitle(title)) {
            throw TitleAlreadyExistsException(title)
        }

        val userId = userProvider.currentUserId()
        val post = Post(
            id = PostId(idGenerator.generate()),
            title = title,
            content = Content(request.content),
            userId = userId
        )

        return repository.save(post)
    }
}
